File: sdk-python/nexus_sdk/graph.py
"""Graph memory operations for Nexus SDK."""

from __future__ import annotations

from typing import Any

import httpx


class Graph:
    """Synchronous graph memory operations."""

    def __init__(self, client: httpx.Client):
        self._client = client

    def create_relationship(
        self,
        source_type: str,
        source_id: str,
        target_type: str,
        target_id: str,
        relationship_type: str,
        weight: float = 1.0,
        metadata: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """
        Create a relationship between two nodes.

        Args:
            source_type: Type of source node (memory, agent, capability)
            source_id: ID of source node
            target_type: Type of target node
            target_id: ID of target node
            relationship_type: Type of relationship (references, derived_from, related_to, etc.)
            weight: Relationship weight (0.0 to 1.0)
            metadata: Additional metadata

        Returns:
            Created relationship
        """
        response = self._client.post(
            "/graph/relationships",
            json={
                "source_type": source_type,
                "source_id": str(source_id),
                "target_type": target_type,
                "target_id": str(target_id),
                "relationship_type": relationship_type,
                "weight": weight,
                "metadata": metadata or {},
            },
        )
        response.raise_for_status()
        return response.json()

    def delete_relationship(self, relationship_id: str) -> bool:
        """Delete a relationship. Returns True if deleted."""
        response = self._client.delete(f"/graph/relationships/{relationship_id}")
        return response.status_code in [200, 204]

    def get_edges(
        self,
        node_type: str,
        node_id: str,
        relationship_type: str | None = None,
        direction: str = "both",
    ) -> list[dict[str, Any]]:
        """
        Get edges for a node.

        Args:
            node_type: Type of node (memory, agent, capability)
            node_id: ID of the node
            relationship_type: Filter by relationship type
            direction: Edge direction (outgoing, incoming, both)

        Returns:
            List of edges
        """
        params: dict[str, Any] = {"direction": direction}
        if relationship_type:
            params["relationship_type"] = relationship_type

        response = self._client.get(
            f"/graph/nodes/{node_type}/{node_id}/edges",
            params=params,
        )
        response.raise_for_status()
        return response.json()["edges"]

    def traverse(
        self,
        start_type: str,
        start_id: str,
        max_depth: int = 2,
        relationship_types: list[str] | None = None,
        target_types: list[str] | None = None,
    ) -> dict[str, Any]:
        """
        Traverse the graph from a starting node.

        Args:
            start_type: Type of starting node
            start_id: ID of starting node
            max_depth: Maximum traversal depth (1-5)
            relationship_types: Filter by relationship types
            target_types: Filter by target node types

        Returns:
            Traversal result with nodes and edges
        """
        response = self._client.post(
            "/graph/traverse",
            json={
                "start_type": start_type,
                "start_id": str(start_id),
                "max_depth": max_depth,
                "relationship_types": relationship_types,
                "target_types": target_types,
            },
        )
        response.raise_for_status()
        return response.json()

    def find_path(
        self,
        from_type: str,
        from_id: str,
        to_type: str,
        to_id: str,
        max_depth: int = 5,
    ) -> dict[str, Any]:
        """
        Find shortest path between two nodes.

        Args:
            from_type: Source node type
            from_id: Source node ID
            to_type: Target node type
            to_id: Target node ID
            max_depth: Maximum path length

        Returns:
            Path result with nodes and edges
        """
        response = self._client.post(
            "/graph/path",
            json={
                "from_type": from_type,
                "from_id": str(from_id),
                "to_type": to_type,
                "to_id": str(to_id),
                "max_depth": max_depth,
            },
        )
        response.raise_for_status()
        return response.json()

    def get_related_memories(
        self,
        memory_id: str,
        relationship_types: list[str] | None = None,
        max_depth: int = 1,
    ) -> list[dict[str, Any]]:
        """
        Get memories related to a given memory.

        Args:
            memory_id: Source memory ID
            relationship_types: Filter by relationship types
            max_depth: How many hops to traverse

        Returns:
            List of related memories with relationship info
        """
        params: dict[str, Any] = {"max_depth": max_depth}
        if relationship_types:
            params["relationship_types"] = relationship_types

        response = self._client.get(
            f"/graph/memories/{memory_id}/related",
            params=params,
        )
        response.raise_for_status()
        return response.json()["memories"]


class GraphAsync:
    """Asynchronous graph memory operations."""

    def __init__(self, client: httpx.AsyncClient):
        self._client = client

    async def create_relationship(
        self,
        source_type: str,
        source_id: str,
        target_type: str,
        target_id: str,
        relationship_type: str,
        weight: float = 1.0,
        metadata: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Create a relationship between two nodes."""
        response = await self._client.post(
            "/graph/relationships",
            json={
                "source_type": source_type,
                "source_id": str(source_id),
                "target_type": target_type,
                "target_id": str(target_id),
                "relationship_type": relationship_type,
                "weight": weight,
                "metadata": metadata or {},
            },
        )
        response.raise_for_status()
        return response.json()

    async def delete_relationship(self, relationship_id: str) -> bool:
        """Delete a relationship."""
        response = await self._client.delete(f"/graph/relationships/{relationship_id}")
        return response.status_code in [200, 204]

    async def get_edges(
        self,
        node_type: str,
        node_id: str,
        relationship_type: str | None = None,
        direction: str = "both",
    ) -> list[dict[str, Any]]:
        """Get edges for a node."""
        params: dict[str, Any] = {"direction": direction}
        if relationship_type:
            params["relationship_type"] = relationship_type

        response = await self._client.get(
            f"/graph/nodes/{node_type}/{node_id}/edges",
            params=params,
        )
        response.raise_for_status()
        return response.json()["edges"]

    async def traverse(
        self,
        start_type: str,
        start_id: str,
        max_depth: int = 2,
        relationship_types: list[str] | None = None,
        target_types: list[str] | None = None,
    ) -> dict[str, Any]:
        """Traverse the graph from a starting node."""
        response = await self._client.post(
            "/graph/traverse",
            json={
                "start_type": start_type,
                "start_id": str(start_id),
                "max_depth": max_depth,
                "relationship_types": relationship_types,
                "target_types": target_types,
            },
        )
        response.raise_for_status()
        return response.json()

    async def find_path(
        self,
        from_type: str,
        from_id: str,
        to_type: str,
        to_id: str,
        max_depth: int = 5,
    ) -> dict[str, Any]:
        """Find shortest path between two nodes."""
        response = await self._client.post(
            "/graph/path",
            json={
                "from_type": from_type,
                "from_id": str(from_id),
                "to_type": to_type,
                "to_id": str(to_id),
                "max_depth": max_depth,
            },
        )
        response.raise_for_status()
        return response.json()

    async def get_related_memories(
        self,
        memory_id: str,
        relationship_types: list[str] | None = None,
        max_depth: int = 1,
    ) -> list[dict[str, Any]]:
        """Get memories related to a given memory."""
        params: dict[str, Any] = {"max_depth": max_depth}
        if relationship_types:
            params["relationship_types"] = relationship_types

        response = await self._client.get(
            f"/graph/memories/{memory_id}/related",
            params=params,
        )
        response.raise_for_status()
        return response.json()["memories"]

File: sdk-typescript/src/graph.ts
/**
 * Graph memory operations
 */

import { AxiosInstance } from 'axios';

export interface Relationship {
  id: string;
  source_type: 'memory' | 'agent' | 'capability';
  source_id: string;
  target_type: 'memory' | 'agent' | 'capability';
  target_id: string;
  relationship_type: string;
  weight: number;
  metadata: Record<string, unknown>;
  created_by_agent_id?: string;
  created_at: string;
}

export interface CreateRelationshipOptions {
  source_type: 'memory' | 'agent' | 'capability';
  source_id: string;
  target_type: 'memory' | 'agent' | 'capability';
  target_id: string;
  relationship_type: string;
  weight?: number;
  metadata?: Record<string, unknown>;
}

export interface TraverseOptions {
  start_type: 'memory' | 'agent' | 'capability';
  start_id: string;
  max_depth?: number;
  relationship_types?: string[];
  target_types?: string[];
}

export interface TraverseResult {
  nodes: Array<{
    type: string;
    id: string;
    depth: number;
  }>;
  edges: Relationship[];
}

export interface PathResult {
  path: Array<{
    type: string;
    id: string;
  }>;
  edges: Relationship[];
  length: number;
}

export interface RelatedMemory {
  memory_id: string;
  relationship_type: string;
  weight: number;
  depth: number;
}

/**
 * Graph memory client for relationship management.
 */
export class Graph {
  constructor(private client: AxiosInstance) {}

  /**
   * Create a relationship between two nodes.
   */
  async createRelationship(options: CreateRelationshipOptions): Promise<Relationship> {
    const response = await this.client.post<Relationship>('/graph/relationships', {
      source_type: options.source_type,
      source_id: options.source_id,
      target_type: options.target_type,
      target_id: options.target_id,
      relationship_type: options.relationship_type,
      weight: options.weight ?? 1.0,
      metadata: options.metadata ?? {},
    });
    return response.data;
  }

  /**
   * Delete a relationship by ID.
   */
  async deleteRelationship(relationshipId: string): Promise<void> {
    await this.client.delete(`/graph/relationships/${relationshipId}`);
  }

  /**
   * Get edges for a node.
   */
  async getEdges(
    nodeType: 'memory' | 'agent' | 'capability',
    nodeId: string,
    options?: {
      relationship_type?: string;
      direction?: 'outgoing' | 'incoming' | 'both';
    }
  ): Promise<Relationship[]> {
    const params: Record<string, unknown> = {
      direction: options?.direction ?? 'both',
    };
    if (options?.relationship_type) {
      params.relationship_type = options.relationship_type;
    }
    const response = await this.client.get<{ edges: Relationship[] }>(
      `/graph/nodes/${nodeType}/${nodeId}/edges`,
      { params }
    );
    return response.data.edges;
  }

  /**
   * Traverse the graph from a starting node.
   */
  async traverse(options: TraverseOptions): Promise<TraverseResult> {
    const response = await this.client.post<TraverseResult>('/graph/traverse', {
      start_type: options.start_type,
      start_id: options.start_id,
      max_depth: options.max_depth ?? 2,
      relationship_types: options.relationship_types,
      target_types: options.target_types,
    });
    return response.data;
  }

  /**
   * Find shortest path between two nodes.
   */
  async findPath(
    fromType: 'memory' | 'agent' | 'capability',
    fromId: string,
    toType: 'memory' | 'agent' | 'capability',
    toId: string,
    maxDepth: number = 5
  ): Promise<PathResult> {
    const response = await this.client.post<PathResult>('/graph/path', {
      from_type: fromType,
      from_id: fromId,
      to_type: toType,
      to_id: toId,
      max_depth: maxDepth,
    });
    return response.data;
  }

  /**
   * Get memories related to a given memory.
   */
  async getRelatedMemories(
    memoryId: string,
    options?: {
      relationship_types?: string[];
      max_depth?: number;
    }
  ): Promise<RelatedMemory[]> {
    const params: Record<string, unknown> = {
      max_depth: options?.max_depth ?? 1,
    };
    if (options?.relationship_types) {
      params.relationship_types = options.relationship_types;
    }
    const response = await this.client.get<{ memories: RelatedMemory[] }>(
      `/graph/memories/${memoryId}/related`,
      { params }
    );
    return response.data.memories;
  }
}
